src/wcl/doc.h

```diff
@@ -0,0 +1,274 @@
+/*
+ * Copyright 2022 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cassert>
+#include <memory>
+#include <sstream>
+#include <vector>
+
+namespace wcl {
+
+// Internal state for a doc object. It is intentionally not exposed in the
+// public API and should not be used directly.
+struct doc_state {
+  size_t character_count = 0;
+  size_t newline_count = 0;
+  size_t first_width = 0;
+  size_t last_width = 0;
+  size_t max_width = 0;
+
+  // Returns the merged state of this and other.
+  doc_state merged(doc_state other) {
+    doc_state merged;
+
+    merged.character_count = character_count + other.character_count;
+    merged.newline_count = newline_count + other.newline_count;
+
+    merged.last_width = other.last_width;
+    // If other doesn't have any newlines, then the left side needs to be accounted for
+    if (other.newline_count == 0) {
+      merged.last_width += last_width;
+    }
+
+    merged.first_width = first_width;
+    // If the left side doesn't have any newlines, then other side needs to be accounted for
+    if (newline_count == 0) {
+      merged.first_width += other.first_width;
+    }
+
+    merged.max_width = std::max(max_width, other.max_width);
+    // It is possible that the join point is a longer line than any other in either doc
+    merged.max_width = std::max(merged.max_width, last_width + other.first_width);
```

Comment:
yep this was the only edge case, let me make sure you added a test case for it and I think we're kosher